test/Microsoft.AspNetCore.Server.KestrelTests/PipelineExtensionTests.cs

```diff
@@ -0,0 +1,149 @@
+            var writer = _pipe.Writer.Alloc();
+            writer.WriteNumeric(number);
+            await writer.FlushAsync();
+
+            var reader = await _pipe.Reader.ReadAsync();
+            var numAsStr = number.ToString();
+            var expected = Encoding.ASCII.GetBytes(numAsStr);
+            AssertEx.Equal(expected, reader.Buffer.Slice(0, numAsStr.Length).ToArray());
+        }
+
+        [Theory]
+        [InlineData(1)]
+        [InlineData(_ulongMaxValueLength / 2)]
+        [InlineData(_ulongMaxValueLength - 1)]
+        public void WritesNumericAcross(int gapSize)
+        {
+            var writer = _pipe.Writer.Alloc(100);
+            // almost fill up the first block
+            var spacer = new Span<byte>(new byte[writer.Buffer.Length - gapSize]);
+            writer.Write(spacer);
+
+            var bufferLength = writer.Buffer.Length;
+            writer.WriteNumeric(ulong.MaxValue);
+            Assert.NotEqual(bufferLength, writer.Buffer.Length);
+
+            writer.FlushAsync().GetAwaiter().GetResult();
+
+            var reader = _pipe.Reader.ReadAsync().GetAwaiter().GetResult();
+            var numAsString = ulong.MaxValue.ToString();
+            var written = reader.Buffer.Slice(spacer.Length, numAsString.Length);
+            Assert.False(written.IsSingleSpan, "The buffer should cross spans");
+            AssertEx.Equal(Encoding.ASCII.GetBytes(numAsString), written.ToArray());
+        }
+
+        [Theory]
+        [InlineData("\0abcxyz", new byte[] { 0, 97, 98, 99, 120, 121, 122 })]
+        [InlineData("!#$%i", new byte[] { 33, 35, 36, 37, 105 })]
+        [InlineData("!#$%", new byte[] { 33, 35, 36, 37 })]
+        [InlineData("!#$", new byte[] { 33, 35, 36 })]
+        [InlineData("!#", new byte[] { 33, 35 })]
+        [InlineData("!", new byte[] { 33 })]
+        // null or empty
+        [InlineData("", new byte[0])]
+        [InlineData(default(string), new byte[0])]
+        public async Task EncodesAsAscii(string input, byte[] expected)
+        {
+            var reader = await WriteAscii(input);
+            if (expected.Length > 0)
+            {
+                AssertEx.Equal(
+                    expected,
+                    reader.Buffer.ToArray());
+            }
+            else
+            {
+                Assert.Equal(0, reader.Buffer.Length);
+            }
+        }
+
+        [Theory]
+        // non-ascii characters stored in 32 bits
+        [InlineData("𤭢𐐝")]
+        // non-ascii characters stored in 16 bits
+        [InlineData("ñ٢⛄⛵")]
+        public async Task WriteAsciiWritesOnlyOneBytePerChar(string input)
+        {
+            // WriteAscii doesn't validate if characters are in the ASCII range
+            // but it shouldn't produce more than one byte per character
+            var reader = await WriteAscii(input);
+            Assert.Equal(input.Length, reader.Buffer.Length);
+        }
+
+        private async Task<ReadResult> WriteAscii(string input)
```

Comment:
Call it `WriteAsciiNoValidation` too, otherwise it's a little confusing to read.